### DISCUSSION TITLE
CustomException 관련 논의

### DISCUSSION THREAD

**Discussion by her0807** [woowacourse-teams/2022-ternoko#85] [Show and tell]
#56

CustomException 을 만드는 기준에 관하여 논의 드리려고 합니다. 

현재는` RunTimeException을 상속받는 Http 응답 관련 예외 클래스`를 만들고  (BadRequestException, NotFountException .. 등등) 

그 하위에 도메인 별로 해당 클래스를 상속받은 형태로 예외 클래스들이 늘어나고 있습니다.  😂
요구사항을 구현하다보면 점점 더 늘어날 것 같은데요. 

<img width="348" alt="스크린샷 2022-07-21 오전 9 23 19" src="https://user-images.githubusercontent.com/26570275/180104293-16a14dd6-5004-49dc-9acc-1264444de7cb.png">

### 따라서 예외가 추가 될 때마다 클래스가 늘어나지 않는 구조를 고민해봐야 할 것 같습니다. 🥲

현재 생각해본 방법으로는 

1. `RunTimeException` 을 상속받는 **Http 응답 관련 예외클래스**를 만들고 `code`를 상태로, 생성자를 `ExceptionType`으로 생성되게 둡니다.

<img width="610" alt="스크린샷 2022-07-21 오전 9 26 42" src="https://user-images.githubusercontent.com/26570275/180104571-dea366ee-63ce-4a6e-876c-1bd2e4ab9c8a.png">

2. 그 하위에 도메인 객체 별 응답 클래스를 만듭니다.

```java

class InvalidCoachException
class InvalidCrewException
class InvalidMemberException
class InvalidReservationException

```

3. 클래스로 늘어날 법한 exception 을 ExceptionType 으로 선언하고, 해당 code 와 message로 예외를 구분합니다.

<img width="687" alt="스크린샷 2022-07-21 오전 9 18 03" src="https://user-images.githubusercontent.com/26570275/180103846-02649103-b35e-41de-a4fb-dcd7b8b9236b.png">

기존 같으면, **BadRequestException** 을 상속받은 **CoachBadRequestException .. CoachNotFoundException ..** 등등 `한 도메인에 대해서 httpException 만큼 늘어나는데, `

이걸 도메인당 1개씩 도메인_InvalidException 을 만들어주고, 해당 내용은 `ExceptionType` 으로 구분해주자! 
이 뜻 인거죠! 어떠신가요 ?

<img width="697" alt="스크린샷 2022-07-21 오전 9 34 07" src="https://user-images.githubusercontent.com/26570275/180105159-aa02ddc8-b2d0-4036-8c28-2be3290d02ba.png">

**Comment by her0807** [comment]
@HyeonbinSa @Juhyung990122 @dongho108  @soominsohn  의견 부탁드립니다!

**Comment by her0807** [comment]
## FE 에서 예외 처리를 할 때 
주로 response statusCode 를 보나요? response ErrorMessage 를 보나요 ? 🤔 @lokba  @sanaandmomo

**Comment by dongho108** [comment]
예외처리구조에 대한 깊은 고민이 담긴 글 감사합니다!!! 예외처리 신경써주셔서 너무감사해요!!

어제 코치의 되는시간이 아닌 시간에 면담을 신청했을 때의 예외처리를 한 적이 있었습니다!

저는 `INVALID_RESERVATION_DATE()`안에 메세지만 `"코치의 되는시간이 아닙니다."` 를 추가해주고 싶었는데 메세지마다 인스턴스가 고정되어있어서 인스턴스를 하나 더 만들게 되더라구요!

아마 이 구조면 메세지마다 인스턴스가 늘어나는 구조가 될 것 같아요!

메세지마다 인스턴스가 늘어나는 구조가 잘못되었는지는 모르겠고 오히려 예외를 세분화 시킬 수 있다는 생각도 듭니다.
그런데 메세지만 약간 다르고 같은 예외라면 메시지만 바꿀 수 있는 방법이 있다면 더 좋겠다는 생각도 듭니다! 😄

**Comment by Juhyung990122** [comment]
수달 고민하시느라 고생하셨습니다!
상속구조를 활용하여 예외를 관리하는 것은 좋은 생각인 것 같습니다ㅎㅎ
체계적으로 관리할 수 있어서 장기적으로 보았을때 좋네요.

다만 메세지는 파라미터로 입력받고, 에러코드는 예외 내에서 관리하는건 어떨까요?
```java 
public class InvalidReservationDateException extends BadRequestException {

    public InvalidReservationDateException(final String message) {
        super(400, message);
    }
}

new InvalidReservationDateException("예외메세지");
```
이유를 말씀드리자면 저는 
- Enum이 갯수와 종류가 명확히 정해져있는 요소에 대해서 적용해야하는 클래스 타입이라고 생각함
- 예외가 늘어날때마다 EnumType에 추가시켜주어야하는 점
- 예외 내용을 파악하기 위해선 EnumType으로 한번 더 들어가서 확인해야한다는 점
- EnumType에 명시한다하더라도 재활용이 어렵다는 점
때문에 예외코드는 숫자로, 예외 메세지는 스트링으로만 관리해도 좋을 것 같아요!

**Reply by Juhyung990122** [reply]
enum 대안으로는 EnumType을 public class로 바꾸고 예외를 다루는 객체(CustomExceptionForTernoko)를 하나 만들어서 
private static final COACH_NOT_FOUND = new CustomExceptionForTernoko(400, "메세지");
이런식으로 관리해도 좋을 것 같네요!

**Comment by HyeonbinSa** [comment]
현재 프로젝트 구조를 봤을 때 Custom Exception 클래스가 엄청나게 많이 생성되지는 않을 것으로 보입니다.
하지만 도메인 별로 나눠서 관리하는 방법을 보니 `도메인`을 기준으로 생성할 수 있어서 개수에 관련된 고민을 해결할 수 있겠다는 생각이 들었습니다.
여기서 생기는 고민은 BadReqeustException을 도메인 별로 나눈 것처럼 `UnAuthrizedException`,  또 새롭게 생겨나는 예외 상황도 도메인 별로 클래스를 생성해야하는가입니다. 그렇게 된다면 이전과 비슷한 개수의 클래스를 관리하게 될 것 같습니다.

그래서 저는 현재 사용하는 구조를 따라가되 exception 패키지에 하위 패키지로 상속 관계를 묶어 관리하는 방안으로 가면 좋을 것 같습니다.

**Comment by soominsohn** [comment]
수달! 깊은 고민의 흔적이 느껴지는 discussion이네요 ㅎㅎ !
상속구조로 exception 클래스도 나누고, exception클래스 안에 필드로 status코드와 exceptionMessage를 가져가니 controllerAdvice에서 처리도 쉽게 가능하겠군요!
다만 enum에서 비슷하지만, 디테일한 상황이 다른 에러에 대해 명명을 어떻게 할지에 대한 고민만 더 해보면 좋을 것 같아요! ㅎㅎ
